tests/test_metrics.py

```diff
@@ -190,6 +195,114 @@ def testValidEmpty(self):
+        config.connections.labelEnd = nameEnd
+        config.targetStart = nameStart + ".run"
+        config.targetEnd = nameEnd + ".run"
+        config.connections.package = "ap_pipe"
+        config.connections.metric = "DummyTime"
+        return config
+
+    @classmethod
+    def makeTask(cls):
+        return PipelineTimingMetricTask(config=cls._makeConfig(nameStart="first", nameEnd="last"))
+
+    def setUp(self):
+        super().setUp()
+        self.metric = Name("ap_pipe.DummyTime")
+
+        self.startTask = DummyTask(name="first")
+        self.startTask.run()
+        self.endTask = DummyTask(name="last")
+        self.endTask.run()
+
+    def testSingleTask(self):
+        task = PipelineTimingMetricTask(config=self._makeConfig(nameStart="first", nameEnd="first"))
+
+        altConfig = lsst.verify.tasks.TimingMetricConfig()
+        altConfig.connections.labelName = "first"
+        altConfig.target = "first.run"
+        altConfig.connections.package = "verify"
+        altConfig.connections.metric = "DummyTime"
+        altTask = lsst.verify.tasks.TimingMetricTask(config=altConfig)
+
+        result = task.run(self.startTask.getFullMetadata(), self.startTask.getFullMetadata())
+        oracle = altTask.run(self.startTask.getFullMetadata())
+
+        self.assertEqual(result.measurement.metric_name, self.metric)
+        self.assertAlmostEqual(result.measurement.quantity.to_value(u.s),
+                               oracle.measurement.quantity.to_value(u.s))
+
+    def testTwoTasks(self):
+        firstTask = PipelineTimingMetricTask(config=self._makeConfig(nameStart="first", nameEnd="first"))
+        secondTask = PipelineTimingMetricTask(config=self._makeConfig(nameStart="last", nameEnd="last"))
+
+        result = self.task.run(self.startTask.getFullMetadata(), self.endTask.getFullMetadata())
+        firstResult = firstTask.run(self.startTask.getFullMetadata(), self.startTask.getFullMetadata())
+        secondResult = secondTask.run(self.endTask.getFullMetadata(), self.endTask.getFullMetadata())
+
+        self.assertEqual(result.measurement.metric_name, self.metric)
+        self.assertGreater(result.measurement.quantity, 0.0 * u.s)
+        self.assertGreaterEqual(result.measurement.quantity,
+                                firstResult.measurement.quantity + secondResult.measurement.quantity)
+
+    def testRunDifferentMethodFirst(self):
+        config = self._makeConfig(nameStart="first", nameEnd="last")
+        config.targetStart = "first.doProcess"
+        task = PipelineTimingMetricTask(config=config)
+        try:
+            result = task.run(self.startTask.getFullMetadata(), self.endTask.getFullMetadata())
+        except lsst.pipe.base.NoWorkFound:
+            # Correct behavior
+            pass
+        else:
+            # Alternative correct behavior
+            testUtils.assertValidOutput(task, result)
+            meas = result.measurement
+            self.assertIsNone(meas)
```

Comment:
This really confuses me: how can the `else` here be valid, since I read `targetStart = "first.doProcess"` as forcing a `NoWorkFound`. Similarly below.

Maybe some docstrings on these two tests would help clarify the intended behavior?

Reply:
[The spec on `MetricTask` says that both of these are allowed](https://pipelines.lsst.io/v/daily/modules/lsst.verify/tasks/lsst.verify.tasks.MetricTask.html#error-handling). I do not write tests that restate the implementation.

Reply:
I guess I'm confused about the test behavior, since only one of those two cases is true in the class as-written. Or is the idea that it could be re-implemented with the other behavior and you don't want the test to have to change?

Reply:
The idea is that the test should not depend on the implementation _at all_; it should be expressed entirely in terms of correct (or at least acceptable) behavior. Either the class works as advertised, or it doesn't.

Customizing the "test" to match the internal code, and having it mirror any changes, would mean it's no longer testing anything; it would be like reviewing your own code and saying "Yes, I agree with me!"